substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/PointsToAnalysis.java

```diff
@@ -703,6 +706,63 @@ public HostVM getHostVM() {
         return hostVM;
     }
 
+    /**
+     * Iterate until analysis reaches a fixpoint.
+     *
+     * @param debug debug context
+     * @param duringAnalysisAction  The action taken during analysis. It could be the {@link org.graalvm.nativeimage.hosted.Feature#duringAnalysis}
+     *                              for Native Image generation, but also can be other similar actions for standalone pointsto
+     *                              analysis that does not dependent on {@link com.oracle.svm.hosted.FeatureImpl.DuringAnalysisAccessImpl}
+     *                              and {@link com.oracle.svm.hosted.NativeImageGenerator.AnalysisFeatureHandler}.
+     * @param stopIterationCriteria The criteria to stop the analysis iteration.
+     * @param reportError Error reporting function
+     * @throws InterruptedException
+     */
+    @SuppressWarnings("try")
+    public void iterativelyAnalyze(DebugContext debug, Consumer<AnalysisUniverse> duringAnalysisAction,
+                                   BooleanSupplier stopIterationCriteria, Consumer<String> reportError) throws InterruptedException {
+        int numIterations = 0;
+        while (true) {
+            try (Indent indent2 = debug.logAndIndent("new analysis iteration")) {
+                /*
+                 * Do the analysis (which itself is done in a similar iterative process)
+                 */
+                boolean analysisChanged = finish();
+
+                numIterations++;
+                if (numIterations > 1000) {
+                    /*
+                     * Usually there are < 10 iterations. If we have so many iterations, we
+                     * probably have an endless loop (but at least we have a performance
+                     * problem because we re-start the analysis so often).
+                     */
+                    reportError.accept(String.format("Static analysis did not reach a fix point after %d iterations because a Feature keeps requesting new analysis iterations. " +
+                                    "The analysis itself %s find a change in type states in the last iteration.",
+                            numIterations, analysisChanged ? "DID" : "DID NOT"));
+
+                }
+
+                /*
+                 * Allow features to change the universe.
+                 */
+                try (StopTimer t2 = getProcessFeaturesTimer().start()) {
+                    int numTypes = universe.getTypes().size();
+                    int numMethods = universe.getMethods().size();
+                    int numFields = universe.getFields().size();
+                    duringAnalysisAction.accept(universe);
+
+                    if (stopIterationCriteria.getAsBoolean()) {
+                        if (numTypes != universe.getTypes().size() || numMethods != universe.getMethods().size() || numFields != universe.getFields().size()) {
+                            reportError.accept(
```

Comment:
I think this error reporting code should remain in `NativeImageGenerator`. That avoids passing in a `reportError` handler. And `duringAnalyssiAction` and `stopIterationCriteria` can be combined to a single `Predicate` (i.e., boolean function). That way only one lambda needs to be passed into this method, making it maintainable.

Reply:
Fixed as suggested.